{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8c8f6e12",
   "metadata": {},
   "source": [
    "# SMRT-Tag\n",
    "#### 08/03/2022\n",
    "This notebook demonstrates how to process SMRT-Tag data, and perform the set of analyses described in our associated manuscript. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7969939f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ff3b7f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get path of notebook \n",
    "NOTEBOOK_PATH = os.getcwd()\n",
    "## Set working directory paths\n",
    "BASE_DIR=os.path.join(os.path.dirname(NOTEBOOK_PATH))\n",
    "TOP_DIR=os.path.join(BASE_DIR,'smrt_tag')\n",
    "## Specify the top-level directory \n",
    "TOP_DIR='/Users/snanda/storage/lab/projects/SMRT-Tag/smrt_tag/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff22ac88",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create directory structure \n",
    "directories=[\n",
    " 'ref',\n",
    " 'ref/GRCh37/',\n",
    " 'ref/GRCh38/',\n",
    " 'raw/HG/',\n",
    " 'preprocess/align',\n",
    " 'analyses/HG/demultiplex_genotype/calling_regions',\n",
    " 'analyses/HG/demultiplex_genotype/calling_regions/unique',\n",
    " 'analyses/HG/demultiplex_genotype/align/',\n",
    " 'analyses/HG/demultiplex_genotype/align/mosdepth/',\n",
    " 'analyses/HG/demultiplex_genotype/plp/',\n",
    " 'analyses/HG/demultiplex_genotype/plp_cmp/',\n",
    " 'analyses/HG/variant_calling/',\n",
    " 'analyses/HG/variant_calling/align',\n",
    " 'analyses/HG/variant_calling/downsample',\n",
    " 'analyses/HG/variant_calling/deepvariant'\n",
    "]\n",
    "\n",
    "for directory in directories:\n",
    "    os.makedirs(\"{}/{}\".format(TOP_DIR,directory),exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbf7e263",
   "metadata": {},
   "source": [
    "### 1.) Download reference datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b4504d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#!/usr/bin/env bash\n",
    "\n",
    "## hg38 reference genom\n",
    "    \n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/genome-stratifications/v3.0/v3.0-stratifications-GRCh37.tar.gz \\\n",
    "    -O \"$TOP_DIR/ref/v3.0-stratifications-GRCh37.tar.gz\" && \\\n",
    "    tar -xf \"$TOP_DIR/ref/v3.0-stratifications-GRCh37.tar.gz\" -C \"$TOP_DIR/ref/\"\n",
    "\n",
    "## GRCh37 reference genome\n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/references/GRCh37/hs37d5.fa.gz \\\n",
    "    -O \"$TOP_DIR/ref/GRCh37/hs37d5.fa.gz\" && \\\n",
    "    gunzip -c \"$TOP_DIR/ref/GRCh37/hs37d5.fa.gz\" > \"$TOP_DIR/ref/GRCh37/hs37d5.fa\"\n",
    "\n",
    "## GRCh38 reference genome:\n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/references/GRCh38/GCA_000001405.15_GRCh38_no_alt_analysis_set.fasta.gz \\\n",
    "    -O \"$TOP_DIR/ref/GRCh38/hg38.fa.gz\" && \\\n",
    "        gunzip -c \"$TOP_DIR/ref/GRCh38/hg38.fa.gz\" > \"$TOP_DIR/ref/GRCh38/hg38.fa\"\n",
    "\n",
    "## HG002 \n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/AshkenazimTrio/HG002_NA24385_son/NISTv4.2.1/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz \\\n",
    "    -O \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\"\n",
    "    \n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/AshkenazimTrio/HG002_NA24385_son/NISTv4.2.1/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed \\\n",
    "    -O \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \n",
    "    \n",
    "## HG003\n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/AshkenazimTrio/HG003_NA24149_father/NISTv4.2.1/GRCh37/HG003_GRCh37_1_22_v4.2.1_benchmark.vcf.gz \\\n",
    "    -O \"$TOP_DIR/ref/GRCh37/HG003_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\"\n",
    "    \n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/AshkenazimTrio/HG003_NA24149_father/NISTv4.2.1/GRCh37/HG003_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed \\\n",
    "    -O \"$TOP_DIR/ref/GRCh37/HG003_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed \"\n",
    "        \n",
    "## HG004\n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/AshkenazimTrio/HG004_NA24143_mother/NISTv4.2.1/GRCh37/HG004_GRCh37_1_22_v4.2.1_benchmark.vcf.gz \\\n",
    "    -O \"$TOP_DIR/ref/GRCh37/HG004_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\"\n",
    "    \n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/AshkenazimTrio/HG004_NA24143_mother/NISTv4.2.1/GRCh37/HG004_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed \\\n",
    "    -O \"$TOP_DIR/ref/GRCh37/HG004_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\"\n",
    "        \n",
    "## PacBio Sequel II 11kb reads HG002\n",
    "wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/HG002_NA24385_son/PacBio_SequelII_CCS_11kb/HG002.SequelII.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.bam \\\n",
    "    -O \"$TOP_DIR/ref/HG002-ctrl.SequelII.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.bam\"\n",
    "\n",
    "## HG002 CpG bismark processed data\n",
    "wget https://ont-open-data.s3.amazonaws.com/gm24385_mod_2021.09/bisulphite/cpg/CpG.gz.bismark.zero.cov.gz \\\n",
    "     -O \"$TOP_DIR/ref/CpG.gz.bismark.zero.cov.gz\"\n",
    "     \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3bcd6f7",
   "metadata": {},
   "source": [
    "### 2.) Download SRA and GEO datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c804e31",
   "metadata": {},
   "source": [
    "#### 3a.) Merge demultiplexed BAM files together by condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82a4a0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#!/usr/bin/env bash\n",
    "# merge_HG_ccs_data.sh: bash script to merge HHG{002,003,004} SMRT-Tag ccs bam files by condition\n",
    "# Usage: ./merge_HG_ccs_data.sh\n",
    "set -eo pipefail\n",
    "\n",
    "## HG002\n",
    "pbmerge -j 15 \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG002_PR28--HG002_PR28.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG002_PR29--HG002_PR29.bam\" \\\n",
    "    -o \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG002.bam\"\n",
    "\n",
    "## HG003\n",
    "pbmerge -j 15 \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG003_PR33--HG003_PR33.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG003_PR34--HG003_PR34.bam\" \\\n",
    "    -o \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG003.bam\"\n",
    "\n",
    "## HG004\n",
    "pbmerge -j 15 \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG004_PR30--HG004_PR30.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG004_PR31--HG004_PR31.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG004_PR32--HG004_PR32.bam\" \\\n",
    "    -o \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG004.bam\"\n",
    "    \n",
    "## All HG002 bams\n",
    "pbmerge -j 15 \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A08.demux2.HG002_PR27--HG002_PR27.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A08.demux2.HG002_PR28--HG002_PR28.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A09.demux2.HG002_PR12--HG002_PR12.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A09.demux2.HG002_PR27--HG002_PR27.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A09.demux2.HG002_PR29--HG002_PR29.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A09.demux2.HG002_PR30--HG002_PR30.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A09.demux2.HG002_PR31--HG002_PR31.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A09.demux2.HG002_PR32--HG002_PR32.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG002_PR28--HG002_PR28.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG002_PR29--HG002_PR29.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR27--HG002_PR27.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR28--HG002_PR28.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR29--HG002_PR29.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR30--HG002_PR30.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR31--HG002_PR31.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR32--HG002_PR32.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR33--HG002_PR33.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A17.demux2.HG002_PR34--HG002_PR34.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR35--HG002_PR35.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR36--HG002_PR36.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR37--HG002_PR37.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR38--HG002_PR38.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR39--HG002_PR39.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR40--HG002_PR40.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR41--HG002_PR41.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.long.demux2.HG002_PR42--HG002_PR42.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR35--HG002_PR35.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR36--HG002_PR36.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR37--HG002_PR37.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR38--HG002_PR38.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR39--HG002_PR39.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR40--HG002_PR40.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR41--HG002_PR41.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A18.short.demux2.HG002_PR42--HG002_PR42.bam\" \\\n",
    "    -o \"${TOP_DIR}/raw/HG/SMRT-Tag.demux2.HG002.bam\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "570b71f4",
   "metadata": {},
   "source": [
    "#### 3b.) Run primrose on HG002 bam data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88552a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#!/usr/bin/env bash\n",
    "# run_primrose_HG002_ccs_data.sh: bash script to run primrose on merged HG002 SMRT-Tag ccs data\n",
    "# Usage: ./run_primrose_HG002_ccs_data.sh\n",
    "set -eo pipefail\n",
    "primrose \\\n",
    "    -j 30 \\\n",
    "    --log-level INFO \\\n",
    "    --keep-kinetics \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.demux2.HG002.bam\" \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.demux2.HG002.primrose.bam\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16f9404c",
   "metadata": {},
   "source": [
    "#### 3c.) Align all files to the reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f87cd74",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#!/usr/bin/env bash\n",
    "# align_HG_ccs_data.sh: bash script to align all HG{002,003,004} SMRT-Tag ccs data\n",
    "# Usage: ./align_HG_ccs_data.sh\n",
    "\n",
    "for ccs_file in $(ls \"${TOP_DIR}/raw/HG/*HG*bam\");\n",
    "do;\n",
    "    file=$(basename $ccs_file)\n",
    "    outfile=\"${file/--*/.bam}\"\n",
    "    \n",
    "    pbmm2 align \\\n",
    "    --preset CCS \\\n",
    "    --sort -j 30 \\\n",
    "    --bam-index BAI \\\n",
    "    --log-level DEBUG \\\n",
    "    $ccs_file \\\n",
    "    \"$TOP_DIR/ref/GRCh38/hg38.fa\" \\\n",
    "    \"$TOP_DIR/preprocess/align/${outfile/.bam/.align.sorted.bam}\"\n",
    "    \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2ce1ff0",
   "metadata": {},
   "source": [
    "# Secondary Analyses "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "793b942e",
   "metadata": {},
   "source": [
    "#### 4a.) Extract CpG information for 5mC comparision with bisulfite data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe302f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#!/usr/bin/env bash\n",
    "# 05_cpg2pickle.sh: bash script to run 05_cpg2pickle.py, and extract primrose 5mC predictions per molecule from reads\n",
    "# Usage: ./05_cpg2pickle.sh *primrose aligned BAM*\n",
    "#     --project: Name of project being analyzed\n",
    "#     --output-dir: Directory to save results\n",
    "#\n",
    "## Outputs:\n",
    "#     output-dir/project_cpg_data.pickle: dictionary containing per-molecule 5mC methylation probability from all BAMs\n",
    "\n",
    "$BASE_DIR/samosa_tag/05_cpg2pickle.py \\\n",
    "    \"$TOP_DIR/preprocess/align/SMRT-Tag.demux2.HG002.primrose.align.sorted.bam\" \\\n",
    "    --project HG002 \\\n",
    "    --output-dir \"$TOP_DIR/analyses/HG/\"\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81322871",
   "metadata": {},
   "source": [
    "### 5) Genotype demultiplexing experiment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c3afd00",
   "metadata": {},
   "source": [
    "#### 5a.) Determine private variants for demultiplexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffd2b912",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# demux_process_truth_sets.sh: Process reference regions and variants from HG{002,003,004} to produce shared calling regions and private variants\n",
    "# Usage: ./demux_process_truth_sets.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     HG002_VCF: HG002 GRCh37 benchmark variant calls\n",
    "#     HG003_VCF: HG003 GRCh37 benchmark variant calls\n",
    "#     HG004_VCF: HG004 GRCh37 benchmark variant calls\n",
    "#     HG002_BED: HG002 GRCh37 verified variant calling regions\n",
    "#     HG003_BED: HG003 GRCh37 verified variant calling regions\n",
    "#     HG004_BED: HG004 GRCh37 verified variant calling regions\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to ${TOPDIR}/analyses/HG/demultiplex_genotype/calling_regions/:\n",
    "#         HG002-4.calling_regions.bed.gz: Shared set of GRCh37 verified calling regions\n",
    "#\n",
    "#     Outputs are written to ${TOPDIR}/analyses/HG/demultiplex_genotype/calling_regions/unique/\n",
    "#         unique.HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz: Unique variants for HG002\n",
    "#         unique.HG003_GRCh37_1_22_v4.2.1_benchmark.vcf.gz: Unique variants for HG003\n",
    "#         unique.HG004_GRCh37_1_22_v4.2.1_benchmark.vcf.gz: Unique variants for HG004\n",
    "\n",
    "\n",
    "${BASE_DIR}/demultiplex_genotype/demux_process_truth_sets.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"$TOP_DIR/ref/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/HG003_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/HG004_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"$TOP_DIR/ref/HG003_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"$TOP_DIR/ref/HG004_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f065dc96",
   "metadata": {},
   "source": [
    "#### 5b.) Align BAMs to hs37d5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "402d02d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# pbmm2_align.sh: Align PacBio unaligned BAM files to a reference genome and run mosdepth to calculate depth stats.\n",
    "# Usage: ./pbmm2_align.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     BAM: PacBio BAM file to be aligned\n",
    "#     OUTPREFIX: Prefix to append to output files\n",
    "#     OUTDIR: Directory to write output\n",
    "#     FASTA: Reference genome FASTA file\n",
    "#     STRAT_BED: Region file for depth calculations - GRCh37_notinalllowmapandsegdupregions.bed.gz\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to $OUTDIR\n",
    "#         $OUTDIR/$OUTPREFIX.align.sorted.bam: Aligned BAM file\n",
    "#         $OUTDIR/mosdepth/$OUTPREFIX.mosdepth* : mosdepth files including (per-base, regions, thresholds.bed etc.)\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/pbmm2_align.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG002.bam\" \\\n",
    "    HG002\n",
    "    \"${TOP_DIR}/analyses/HG/demultiplex_genotype/align/\"\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "    \n",
    "$BASE_DIR/variant_calling/pbmm2_align.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG003.bam\" \\\n",
    "    HG003\n",
    "    \"${TOP_DIR}/analyses/HG/demultiplex_genotype/align/\"\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "    \n",
    "$BASE_DIR/variant_calling/pbmm2_align.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.A16.demux2.HG004.bam\" \\\n",
    "    HG004\n",
    "    \"${TOP_DIR}/analyses/HG/demultiplex_genotype/align/\"\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a5a693e",
   "metadata": {},
   "source": [
    "#### 5c.) Naive variant calling in GIAB benchmark regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a76fc779",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# demux_plp2vcf.sh: Naively calls variants using samtools mpileup in the regions provided\n",
    "# Usage: ./demux_plp2vcf.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     BAM: BAM-file to call variants\n",
    "#     FASTA: GRCh37, hs37d5, reference genome fasta\n",
    "#     CALLING_REGIONS: BED file containing shared set of GRCh37 verified calling regions produced by demux_process_truth_sets.sh\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to ${TOPDIR}/analyses/HG/demultiplex_genotype/plp/:\n",
    "#         BAMfile.q15.d2.vcf.gz: Naive VCF containing SNVs called in the provided regions with at least 2 reads and a minimum MAPQ of 15\n",
    "\n",
    "$BASE_DIR/demultiplex_genotype/demux_plp2vcf.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/analyses/HG/demultiplex_genotype/align/HG002.align.sorted.bam\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOPDIR}/analyses/HG//calling_regions/HG002-4.calling_regions.bed.gz\"\n",
    "    \n",
    "$BASE_DIR/demultiplex_genotype/demux_plp2vcf.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/analyses/HG/demultiplex_genotype/align/HG003.align.sorted.bam\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOPDIR}/analyses/HG/demultiplex_genotype/calling_regions/HG002-4.calling_regions.bed.gz\"\n",
    "    \n",
    "$BASE_DIR/demultiplex_genotype/demux_plp2vcf.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/analyses/HG/demultiplex_genotype/align/HG004.align.sorted.bam\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOPDIR}/analyses/HG/demultiplex_genotype/calling_regions/HG002-4.calling_regions.bed.gz\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77d8b7e",
   "metadata": {},
   "source": [
    "#### 5d.) Intersect called variants with private variants to determine degree of shared variants "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7abc8acf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# demux_compare.sh: Intersect called variants with private variants to determine degree of shared variants \n",
    "# Usage: ./demux_compare.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     BAM: BAM-file to call variants\n",
    "#     REGIONS: Region file for calculations - GRCh37_notinalllowmapandsegdupregions.bed.gz\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to ${TOPDIR}/analyses/HG/demultiplex_genotype/plp_cmp/:\n",
    "#         depth2_bed/HG{002,003,004}.d2.GRCh37_notinalldifficultregions.bed.gz: BED file indicating regions where at least 2 reads are present\n",
    "#         HG{002,003,004}.q15.d2_vs_HG{002,003,004}_unique.vcf.gz: VCF file containing variants present in a sequenced sample and private benchmark variants.\n",
    "#         HG{002,003,004}.q15.d2_vs_HG{002,003,004}_unique.stats: Stats for above VCF file\n",
    "#         HG{002,003,004}.d2_vs_HG{002,003,004}.base.vcf.gz: VCF file containing shared variants in experimental regions (at least 2 reads are present)\n",
    "#         HG{002,003,004}.d2_vs_HG{002,003,004}.base.stats: Stats for above VCF file\n",
    "\n",
    "$BASE_DIR/demultiplex_genotype/demux_compare.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"$TOP_DIR/ref/GRCh37/union/GRCh37_notinalldifficultregions.bed.gz\"\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a5ac6d3",
   "metadata": {},
   "source": [
    "### 6.) Variant Calling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d496fb1",
   "metadata": {},
   "source": [
    "#### 6a.) Aligning HG002 reads to the genome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a41151",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# pbmm2_align.sh: Align PacBio unaligned BAM files to a reference genome and run mosdepth to calculate depth stats.\n",
    "# Usage: ./pbmm2_align.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     BAM: PacBio BAM file to be aligned\n",
    "#     OUTPREFIX: Prefix to append to output files\n",
    "#     OUTDIR: Directory to write output\n",
    "#     FASTA: Reference genome FASTA file\n",
    "#     REGIONS: Region file for depth calculations - GRCh37_notinalllowmapandsegdupregions.bed.gz\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to $OUTDIR\n",
    "#         $OUTDIR/$OUTPREFIX.align.sorted.bam: Aligned BAM file\n",
    "#         $OUTDIR/mosdepth/$OUTPREFIX.mosdepth* : mosdepth files including (per-base, regions, thresholds.bed etc.)\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/pbmm2_align.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/raw/HG/SMRT-Tag.demux2.HG002.bam\" \\\n",
    "    HG002\n",
    "    \"${TOP_DIR}/analyses/HG/variant_calling/align/\"\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a3811cb",
   "metadata": {},
   "source": [
    "#### 6b.) Subsampling aligned BAM files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90e35af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# downsample_bam.sh: Downsample aligned BAM files produced by pbmm2_align.sh\n",
    "# Usage: ./downsample_bam.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     BAM: PacBio BAM file to be aligned\n",
    "#     DEPTH: Depth of downsampling\n",
    "#     FRAC: Fraction provided to samtools for downsampling\n",
    "#     REGIONS: Region file for depth calculations - GRCh37_notinalllowmapandsegdupregions.bed.gz\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to ${TOPDIR}/analyses/HG/variant_calling/downsample/${PREFIX}/${DEPTH}X/seed${SEED}\n",
    "#         $OUTDIR/${PREFIX}.${DEPTH}X.seed${SEED}.bam: Downsampled aligned BAM file\n",
    "#         $OUTDIR/mosdepth/$OUTPREFIX.mosdepth* : mosdepth files including (per-base, regions, thresholds.bed etc.)\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/downsample_bam.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/analyses/HG/variant_calling/align/HG002.align.sorted.bam\" \\\n",
    "    10 \\\n",
    "    0.895 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "    \n",
    "$BASE_DIR/variant_calling/downsample_bam.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/analyses/HG/variant_calling/align/HG002.align.sorted.bam\" \\\n",
    "    5 \\\n",
    "    0.44 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "    \n",
    "    \n",
    "$BASE_DIR/variant_calling/downsample_bam.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOP_DIR}/analyses/HG/variant_calling/align/HG002.align.sorted.bam\" \\\n",
    "    3 \\\n",
    "    0.268 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "\n",
    "\n",
    "\n",
    "## Subsample control data as well\n",
    "$BASE_DIR/variant_calling/downsample_bam.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"$TOP_DIR/ref/HG002-ctrl.SequelII.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.bam\" \\\n",
    "    15 \\\n",
    "    1.0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "\n",
    "$BASE_DIR/variant_calling/downsample_bam.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"$TOP_DIR/ref/HG002-ctrl.SequelII.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.bam\" \\\n",
    "    10 \\\n",
    "    0.66 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "\n",
    "$BASE_DIR/variant_calling/downsample_bam.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"$TOP_DIR/ref/HG002-ctrl.SequelII.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.bam\" \\\n",
    "    5 \\\n",
    "    0.33 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "\n",
    "$BASE_DIR/variant_calling/downsample_bam.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"$TOP_DIR/ref/HG002-ctrl.SequelII.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.bam\" \\\n",
    "    3 \\\n",
    "    0.20 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/union/GRCh37_notinalllowmapandsegdupregions.bed.gz\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "803ddfe0",
   "metadata": {},
   "source": [
    "#### 6c.) Running deepvariant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da88bb15",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# deepvariant.sh: Call variants from PacBio HiFi reads using deepvariant\n",
    "# Usage: ./deepvariant.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     BAM: PacBio BAM file to be \n",
    "#     OUTPREFIX: Prefix to append to output files\n",
    "#     FASTA: Genome FASTA file used for alignment\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to ${TOPDIR}/analyses/HG/variant_calling/deepvariant/$OUTPREFIX\n",
    "#          $OUTDIR/$OUTPREFIX.deepvariant.vcf.gz: Deepvariant VCF\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/deepvariant.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOPDIR}/analyses/HG/variant_calling/downsample/HG002/10X/seed0/HG002.10X.seed0.bam\" \\\n",
    "    HG002.10X.seed0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\"\n",
    "    \n",
    "    \n",
    "    \n",
    "$BASE_DIR/variant_calling/deepvariant.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOPDIR}/analyses/HG/variant_calling/downsample/HG002/5X/seed0/HG002.5X.seed0.bam\" \\\n",
    "    HG002.5X.seed0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\"\n",
    "    \n",
    "    \n",
    "$BASE_DIR/variant_calling/deepvariant.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOPDIR}/analyses/HG/variant_calling/downsample/HG002/3X/seed0/HG002.3X.seed0.bam\" \\\n",
    "    HG002.3X.seed0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\"\n",
    "    \n",
    "    \n",
    "    \n",
    "$BASE_DIR/variant_calling/deepvariant.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOPDIR}/analyses/HG/variant_calling/downsample/HG002-ctrl/15X/seed0/HG002-ctrl.15X.seed0.bam\" \\\n",
    "    HG002-ctrl.15X.seed0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\"\n",
    "    \n",
    "    \n",
    "$BASE_DIR/variant_calling/deepvariant.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOPDIR}/analyses/HG/variant_calling/downsample/HG002-ctrl/10X/seed0/HG002-ctrl.10X.seed0.bam\" \\\n",
    "    HG002-ctrl.10X.seed0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\"\n",
    "    \n",
    "    \n",
    "$BASE_DIR/variant_calling/deepvariant.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOPDIR}/analyses/HG/variant_calling/downsample/HG002-ctrl/5X/seed0/HG002-ctrl.5X.seed0.bam\" \\\n",
    "    HG002-ctrl.5X.seed0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\"\n",
    "    \n",
    "\n",
    "$BASE_DIR/variant_calling/deepvariant.sh \\\n",
    "    $TOP_DIR \\\n",
    "    \"${TOPDIR}/analyses/HG/variant_calling/downsample/HG002-ctrl/3X/seed0/HG002-ctrl.3X.seed0.bam\" \\\n",
    "    HG002-ctrl.3X.seed0 \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\"\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d847614",
   "metadata": {},
   "source": [
    "#### 6d.) Evaluate deepvariant variant calls with hap.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0859d926",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#! /usr/bin/env bash\n",
    "# happy.sh: Benchmark vcf against truth set using hap.py\n",
    "# Usage: ./happy.sh \n",
    "#\n",
    "## Inputs:\n",
    "#     TOPDIR: $TOP_DIR\n",
    "#     VCF: Deepvariant VCF file\n",
    "#     TRUTH_VCF: Truth VCF file from GIAB\n",
    "#     TRUTH_BED: Truth BED file from GIAB\n",
    "#     FASTA: Genome FASTA file used for alignment\n",
    "#     STRAT: Genome stratifications for GRCh37 from GIAB - v3.0-GRCh37-v4.2.1-stratifications.tsv\n",
    "#\n",
    "## Outputs:\n",
    "#     Outputs are written to ${TOPDIR}/analyses/HG/variant_calling/deepvariant/$OUTPREFIX:\n",
    "#         $OUTDIR/$OUTPREFIX.deepvariant.happy: hap.py output file benchmarking variant calling\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/happy.sh \\\n",
    "    $TOP_DIR \\\n",
    "    ${TOPDIR}/analyses/HG/variant_calling/deepvariant/HG002.10X.seed0/HG002.10X.seed0.deepvariant.vcf.gz \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/v3.0-GRCh37-v4.2.1-stratifications.tsv\"\n",
    "\n",
    "$BASE_DIR/variant_calling/happy.sh \\\n",
    "    $TOP_DIR \\\n",
    "    ${TOPDIR}/analyses/HG/variant_calling/deepvariant/HG002.5X.seed0/HG002.5X.seed0.deepvariant.vcf.gz \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/v3.0-GRCh37-v4.2.1-stratifications.tsv\"\n",
    "\n",
    "$BASE_DIR/variant_calling/happy.sh \\\n",
    "    $TOP_DIR \\\n",
    "    ${TOPDIR}/analyses/HG/variant_calling/deepvariant/HG002.3X.seed0/HG002.3X.seed0.deepvariant.vcf.gz \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/v3.0-GRCh37-v4.2.1-stratifications.tsv\"\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/happy.sh \\\n",
    "    $TOP_DIR \\\n",
    "    ${TOPDIR}/analyses/HG/variant_calling/deepvariant/HG002-ctrl.15X.seed0/HG002-ctrl.15X.seed0.deepvariant.vcf.gz \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/v3.0-GRCh37-v4.2.1-stratifications.tsv\"\n",
    "\n",
    "$BASE_DIR/variant_calling/happy.sh \\\n",
    "    $TOP_DIR \\\n",
    "    ${TOPDIR}/analyses/HG/variant_calling/deepvariant/HG002-ctrl.10X.seed0/HG002-ctrl.10X.seed0.deepvariant.vcf.gz \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/v3.0-GRCh37-v4.2.1-stratifications.tsv\"\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/happy.sh \\\n",
    "    $TOP_DIR \\\n",
    "    ${TOPDIR}/analyses/HG/variant_calling/deepvariant/HG002-ctrl.5X.seed0/HG002-ctrl.5X.seed0.deepvariant.vcf.gz \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/v3.0-GRCh37-v4.2.1-stratifications.tsv\"\n",
    "\n",
    "\n",
    "$BASE_DIR/variant_calling/happy.sh \\\n",
    "    $TOP_DIR \\\n",
    "    ${TOPDIR}/analyses/HG/variant_calling/deepvariant/HG002-ctrl.3X.seed0/HG002-ctrl.3X.seed0.deepvariant.vcf.gz \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark.vcf.gz\" \\\n",
    "    \"$TOP_DIR/ref/GRCh37/HG002_GRCh37_1_22_v4.2.1_benchmark_noinconsistent.bed\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/hs37d5.fa\" \\\n",
    "    \"${TOP_DIR}/ref/GRCh37/v3.0-GRCh37-v4.2.1-stratifications.tsv\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
